mirgecom/euler.py

```diff
@@ -20,16 +27,195 @@
+    momflux = make_obj_array(
+        [
+            (mom[i] * mom[j] / mass + (p if i == j else 0))
+            for i in range(ndim)
+            for j in range(ndim)
+        ]
+    )
+    massflux = mom * make_obj_array([1.0])
+    energyflux = mom * make_obj_array([(energy + p) / mass])
+
+    flux = flat_obj_array(massflux, energyflux, momflux,)
+
+    return flux
+
+
+def _get_wavespeed(w, eos=IdealSingleGas()):
+
+    mass = w[0]
+    mom = w[2:]
+
+    v = mom * make_obj_array([1.0 / mass])
+
+    sos = eos.sound_speed(w)
+    wavespeed = clmath.sqrt(np.dot(v, v)) + sos
+    return wavespeed
+
+
+def _facial_flux(discr, w_tpair, eos=IdealSingleGas()):
+
+    dim = discr.dim
+
+    mass = w_tpair[0]
+    energy = w_tpair[1]
+    mom = w_tpair[2:]
+
+    normal = with_queue(mass.int.queue, discr.normal(w_tpair.dd))
+
+    # Get inviscid fluxes [rhoV (rhoE + p)V (rhoV.x.V + p*I) ]
+    qint = flat_obj_array(mass.int, energy.int, mom.int)
+    qext = flat_obj_array(mass.ext, energy.ext, mom.ext)
+
+    # - Figure out how to manage grudge branch dependencies
+    #    qjump = flat_obj_array(rho.jump, rhoE.jump, rhoV.jump)
+    qjump = qext - qint
+    flux_int = _inviscid_flux(discr, qint, eos)
+    flux_ext = _inviscid_flux(discr, qext, eos)
+
+    # Lax/Friedrichs/Rusonov after JSH/TW Nodal DG Methods, p. 209
+    flux_jump = (flux_int + flux_ext) * make_obj_array([0.5])
+
+    # wavespeeds = [ wavespeed_int, wavespeed_ext ]
+    wavespeeds = [_get_wavespeed(qint), _get_wavespeed(qext)]
+
+    lam = clarray.maximum(wavespeeds[0], wavespeeds[1])
+    lfr = qjump * make_obj_array([0.5 * lam])
+
+    # Surface fluxes should be inviscid flux .dot. normal
+    # rhoV .dot. normal
+    # (rhoE + p)V  .dot. normal
+    # (rhoV.x.V)_1 .dot. normal
+    # (rhoV.x.V)_2 .dot. normal
+    nflux = flat_obj_array(
+        [
+            np.dot(flux_jump[(i * dim): ((i + 1) * dim)], normal)
+            for i in range(dim + 2)
+        ]
+    )
+
+    # add Lax/Friedrichs term
+    flux_weak = nflux + lfr
+
+    return discr.interp(w_tpair.dd, "all_faces", flux_weak)
+
+
+def inviscid_operator(
+    discr,
+    w,
+    t=0.0,
+    eos=IdealSingleGas(),
+    boundaries={BTAG_ALL: DummyBoundary()},
+):
+    """
+    Returns the RHS of the Euler flow equations:
+    d/dt(Q) = - nabla .dot. F + S
+    where state Q = [ rho rhoE rhoV ]
+          flux F = [ rhoV (rhoE + p)V (rho(V.x.V) + p*I) ]
+          sources S = [mass_src energy_src momentum_src]
+    """
+#    :math: \partial_t Q = - \\nabla \\cdot F
+
+    ndim = discr.dim
+
+    vol_flux = _inviscid_flux(discr, w, eos)
+    dflux = flat_obj_array(
+        [
+            discr.weak_div(vol_flux[(i * ndim): (i + 1) * ndim])
+            for i in range(ndim + 2)
+        ]
+    )
+
+    interior_face_flux = _facial_flux(
+        discr, w_tpair=_interior_trace_pair(discr, w), eos=eos
+    )
+
+    # Domain boundaries
+    domain_boundary_flux = sum(
+        boundaries[btag].get_boundary_flux(
```

Comment:
Classes always harbor a temptation of mutable state. Look up `functools.partial`, which IMO can do most of what you'd want a class for here, but with less typing and clearer information flow.

Reply:
(I.e. what I'm pushing for is for theses things to just be a function.

Reply:
I would like to talk about this more.  I did refactor this a bit to make it closer to what you want (I think). Although I strongly prefer the former construction where the boundary call returns a flux across the boundary instead of just a solution - leaving the caller to construct the flux however they choose.